Mypy

daniel-montanari opened the pull request.

daniel-montanari committed:
ignore vscode

daniel-montanari committed:
Merge remote-tracking branch 'upstream/master' into master

daniel-montanari committed:
Merge remote-tracking branch 'upstream/master' into master

daniel-montanari committed:
add mypy

daniel-montanari committed:
create mypy job

daniel-montanari committed:
rationale comments

daniel-montanari committed:
Don't specify python version, tox to override

daniel-montanari commented:
I don't love how I've handled dependencies, possibly another way to go about it is just ignore 3rd party stuff all together, and globally allow for missing imports. This way the only dependency is mypy. I had a look at what we miss out on by not getting type information from other packages, and currently it looks like it actually makes things worse - we get nothing useful out of knowing the types since our own code isn't typed yet, but we still have to deal with suppressing warnings due to runtime import mechanics confusing mypy.

daniel-montanari committed:
only parse sourcetree

clint-lawrence commented:
Me just playing. Didn't think too hard about it, just wanted to see what kind of progress I could make https://github.com/clint-lawrence/Fixate/tree/mypy-experiments

daniel-montanari commented:
> Me just playing. Didn't think too hard about it, just wanted to see what kind of progress I could make https://github.com/clint-lawrence/Fixate/tree/mypy-experiments
> 
> Not free of errors (~10 left), but I did fix a fair few.

_src\fixate\drivers\lcr\agilent_u1732c.py:95: error: Signature of "frequency" incompatible with supertype "LCR"_ 
- overriding None with int 
_src\fixate\drivers\lcr\agilent_u1732c.py:95: error: Signature of "frequency" incompatible with supertype "LCR"_
- overriding attribute with property (bug in mypy)
Current version (0.971) does not allow overriding attributes with properties.
Fixed here: https://github.com/python/mypy/pull/13475#event-7232466187
This should hopefully be added to the next release
Tested using 0.990+dev from master fixes this.

_src\fixate\drivers\funcgen\rigol_dg1022.py:399: error: Name "amplitude_ch1" already defined on line 391_
- mypy doesn't seem to like if you the property setter is not immediately after the property
https://github.com/python/mypy/issues/1465

_src\fixate\drivers\funcgen\rigol_dg1022.py:449: error: "Callable[[FuncGen], Any]" has no attribute "setter"_
- While it isn't great that mypy doesn't support this, I think defining properties on an abstract parent is odd design, probably should change them to normal attributes, then the overriding class can choose to implement them as properties or not
https://github.com/python/mypy/issues/4644

_src\fixate\drivers\pps\__init__.py:22: error: Incompatible types in assignment (expression has type "BK178X", variable has type "SPD3303X")
src\fixate\drivers\pps\__init__.py:23: error: "SPD3303X" has no attribute "baud_rate"_
- looks like the logic in this function is confusing mypy into thinking that it is reassigning driver due to the shadowed variable name

daniel-montanari commented:
```python
if struct.calcsize("P") == 8:  # 64 bit
    FT_HANDLE = ctypes.c_ulonglong
else:
    FT_HANDLE = ctypes.c_ulong
```
Mypy forces you to either rewrite this as a conditional, or explicitly annotate the types yourself :(

clint-lawrence commented:
> ```python
> if struct.calcsize("P") == 8:  # 64 bit
>     FT_HANDLE = ctypes.c_ulonglong
> else:
>     FT_HANDLE = ctypes.c_ulong
> ```
> 
> Mypy forces you to either rewrite this as a conditional, or explicitly annotate the type as a Union yourself :(

`FH_HANDLE = ctypes.c_void_p` should fix that and is a more appropriate type. It is _not_ a Union!

daniel-montanari committed:
Merge remote-tracking branch 'upstream/master'

daniel-montanari committed:
Merge branch 'master' into mypy

daniel-montanari committed:
update mypy, install fixate into tox mypy env

daniel-montanari committed:
add extra rules for following imports

clint-lawrence reviewed (commented):
Overall, I'm happy to merge. I think you should get buy in from everyone first though. (@John2202W , @jcollins1983 , @christopherwallis , @Jasper-Harvey0 )

The main thing I'm wary off is not having a clear goal stated. For example, that goal could be
- Have all interfaces intended for use in scripts annotated
- Have the entire codebase annotated

The first should probably be the priority. (e.g. we spend more time writing scripts than modifying fixate)

And I've already said this before, but not sure I've committed it in writing - we almost certainly shouldn't try annotating the code as is. There is all kinds of busted stuff that most likely needs to be refactored and/or redesigned before annotation will be useful.

daniel-montanari commented:
I've updated the exclude list since files have changed in the last few months.
I've also added all excluded files to the silent imports list. This means that importing an excluding module won't create errors due to mypy analysing the excluded module when it is imported by another module.
I've also updated mypy to the latest version.

I haven't gone through and changed any of the config rules to non-default. We can merge this first and then update the config later.

daniel-montanari commented on tox.ini:
```diff
@@ -36,4 +36,10 @@ addopts = --ignore=test/manual
+# mypy gives different results if you actually install the stuff before you check it
+# don't cache anything, will be slower, but each run should be clean
+# this also stops weirdness around sharing cache with other instances of mypy
+commands = mypy --no-incremental --config-file=mypy.ini
```
With `--no-incremental` it takes 15 seconds, with cache it takes 2 seconds (but still takes the same amount of time the first time it is run)
If we don't specify `--no-incremental` we at least will want something like `--cache-dir="{envdir}/mypy_cache"` because I was getting really weird errors due to tox's mypy and my local mypy sharing caches

jcollins1983 replied:
mypy is at version 1.3 now, is there any reason not to use the latest over the currently pinned 1.2?

daniel-montanari replied:
No idea why 1.2 was installed over 1.3, will upgrade that.

I think as well this also only installs the dependencies for the base fixate package, I think I should add a run for the gui as well.

daniel-montanari replied:
Gui has not much in it and looks like there are things in there that don't have type stubs. There's a package that addresses some of this PyQt5-stubs. I'll leave everything for now and look into it if we have to do any changes to the gui in the future.

daniel-montanari replied:
> The main thing I'm wary off is not having a clear goal stated. For example, that goal could be
>  - Have all interfaces intended for use in scripts annotated

> And I've already said this before, but not sure I've committed it in writing - we almost certainly shouldn't try annotating the code as is. There is all kinds of busted stuff that most likely needs to be refactored and/or redesigned before annotation will be useful.

After looking at this again I think this is the way to go, the current mypy config was written using default settings and turn off stuff that breaks, with the intent of fixing things up and turning them back on. The only problem is that as we turn on stricter and stricter settings this will cause all sorts of things to pop up.

I think the main priority is to make things easier to use in test scripts.

I'd say the things that would making writing test scripts easiest would be
- drivers
- jig_mapping (needs virtualmux rework)
- checks (John added some docstrings explaining what the checks actually do now, typing these would really only enforce whether these are super generic comparisons or meant for numeric types only)
- ui (not cli/qt specific just the generic interface)


Some things that would hopefully be easy to add in
- common (other than testclasses I had no idea all the other stuff was in here)
- control
- exceptions (really need to change these from BaseException at some point)


Modules that could benefit at some point:
- reporting

Maybe other modules can be enabled partially/fully if people work on them?
A lot of things are hard to work on because unstructured clumps of data get passed between modules, in a lot of modules current states, mypy would probably complain a lot

jcollins1983 commented:
> Overall, I'm happy to merge. I think you should get buy in from everyone first though. (@John2202W , @jcollins1983 , @christopherwallis , @Jasper-Harvey0 )
> 
> The main thing I'm wary off is not having a clear goal stated. For example, that goal could be
> 
> * Have all interfaces intended for use in scripts annotated
> * Have the entire codebase annotated
> 
> The first should probably be the priority. (e.g. we spend more time writing scripts than modifying fixate)
> 
> And I've already said this before, but not sure I've committed it in writing - we almost certainly shouldn't try annotating the code as is. There is all kinds of busted stuff that most likely needs to be refactored and/or redesigned before annotation will be useful.

I'm generally in favour of this, particularly if it means you don't have to go digging into functions that are supposed to be providing a layer of abstraction to work out what they actually need to be provided, so I agree with Clint that the priority should be to annotate all script facing interfaces.

daniel-montanari committed:
update mypy to 1.3

daniel-montanari committed:
add isolated cache back in

John2202W commented:
I'm in favour of gradual introduction of type hinting - haven't played much with mypy but appears a good route.

daniel-montanari commented on mypy.ini:
```diff
@@ -0,0 +1,81 @@
+# the advantage with editing the config rather than suppressing things inline is that there is
+# one location to look at in regards to how mypy behaves, and random parts of the code won't be
+# ignored and forgotten about due to scattered comments that locally disable type checking
+
```
Plan is to follow the guide here https://mypy.readthedocs.io/en/latest/existing_code.html

As a reference with checks for everything in scripts and src/fixate:
Default ruleset ->130 errors
Checking untyped code -> 600 errors
Disallowing untyped code -> 2000 errors

christopherwallis commented on tox.ini:
```diff
@@ -42,4 +42,4 @@ deps = mypy==1.3
 # mypy gives different results if you actually install the stuff before you check it
 # don't cache anything, will be slower, but each run should be clean
```
Does the comment need to be updated with the latest change?

daniel-montanari replied:
done

christopherwallis commented:
I'm definitely supportive of this too. Should (eventually) make a lot of things easier, especially the changes to script interfaces. I haven't tested anything and I'm not completely up to date with fixate or mypy but everything looks good based on my read through of the pull request.
The comment about caching with @daniel-montanari's last change might need updating though.

daniel-montanari committed:
=cache comment
